.gitlab-ci.yml

```diff
@@ -120,6 +120,17 @@ test:
   script:
     - cargo test --verbose --all-features --no-fail-fast --workspace
 
+
+docs:
```

Comment:
why not pass `-D warnings` to have it failed instead?
And why the new job instead of adding this to `publish-docs:` job?

Reply:
> why not pass -D warnings to have it failed instead?

Doesn't work for `cargo doc` unfortunately.

>And why the new job instead of adding this to publish-docs: job?

We want to run this for every PR, so that you know if you e.g. have broken intra-crate links in the doc comments. `publish-docs` is only run after merge.

Reply:
it [works](https://github.com/paritytech/substrate/blob/7e94f01af67c498605eb0206b7fca261f12a278c/.gitlab-ci.yml#L477) with `cargo doc`

Reply:
then it would make sense to move the docks generation from that job to this one, and there just copy the artifacts from this job. To save some CO2 you know.

Reply:
> it [works](https://github.com/paritytech/substrate/blob/7e94f01af67c498605eb0206b7fca261f12a278c/.gitlab-ci.yml#L477) with `cargo doc`

So I had checked exactly that this morning, before I implemented it the way it is now. I just went back and re-checked to uphold my own mental sanity. tl;dr is that `export RUSTFLAGS="-D warnings"` doesn't have any effect on `rustdoc` (the flags for `rustdoc` are under `RUSTDOCFLAGS`, but it also doesn't have an effect to set it here).

I also created a simple `lib.rs` with broken doc comments:
```
/// Test
/// ```rust
/// [`Foobar`](`foo::does::not::exist`)
pub fn main() {}
```
If then setting the env to `-Dwarnings` and executing `cargo doc --verbose` these warnings are displayed, but the exit code is always `0`.

I also checked the documentation for `rustdoc` and its source code, but couldn't find any way to have the tool exit with 1 when it detects warnings. I hope I didn't miss something, also didn't find a possibility via search engines.

Reply:
> then it would make sense to move the docks generation from that job to this one, and there just copy the artifacts from this job. To save some CO2 you know.

That's a good point! I haven't really worked with artifacts before, my understanding is that they are collected automatically here, right?

I added `--all-features` to the docs generation here, to make the flags for both `cargo doc` invocations compatible. Is this already sufficient?

Reply:
so you 
1. [set up](https://github.com/paritytech/substrate/blob/master/.gitlab-ci.yml#L480) the collection
2. [take care](https://github.com/paritytech/substrate/blob/master/.gitlab-ci.yml#L490) of putting them there
3. [ask](https://github.com/paritytech/substrate/blob/master/.gitlab-ci.yml#L681) artifacts from this job
4. [skip git](https://github.com/paritytech/substrate/blob/master/.gitlab-ci.yml#L687)
5. draw the rest of the owl